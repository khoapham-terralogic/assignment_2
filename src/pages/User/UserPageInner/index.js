import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types'
import { useHistory } from 'react-router-dom'
import { ClipSpinner, CustomModal } from '../../../components';
import { editSvg, avatar } from '../../../constants'
import UserForm from '../UserForm';


const UserPageInner = ({
    isAuth,
    logoutUser,
    updateProfile,
    userLoading,
    uploadImage,
    avatarLink,
    local_user
}) => {
    const history = useHistory()
    const [isOpen, setIsOpen] = useState(false)
    const toggle = () => { setIsOpen(!isOpen) }
    const callBack = (data) => { setIsOpen(data) }
    useEffect(() => {
        if (!isAuth) {
            history.push("/auth/login")
        }
    }, [history, isAuth])
    // const local_user = JSON.parse(localStorage.getItem("user"))
    // if (local_user)
    return (
        <div className="user__container__body">
            {userLoading ? <div className="loadingBackground"><ClipSpinner size={45} color="#fff" /></div> : null}
            {/* {user &&
                    <>
                        <CustomModal modalCallback={callBack} uploadImage={uploadImage} isLoading={userLoading} isOpen={isOpen} toggle={toggle} />
                        <div className="row row--single row--header">
                            <div className="col-xs-12 col-md-6 user__container__body__header">
                                <div className="user__container__body__header-img-container">
                                    <img className="user__container__body__header-img" src={`http://api.terralogic.ngrok.io/${user.avatar}`} alt="avatar" />
                                    <img onClick={toggle} className="user__container__body__header-subimg" src={editSvg} alt="edit" />
                                </div>
                                <div className="user__container__body__header-name">{local_user.displayName}</div>
                            </div>
                        </div>
                    </>} */}
            {local_user &&
                <>
                    <CustomModal
                        modalCallback={callBack}
                        uploadImage={uploadImage}
                        isLoading={userLoading}
                        isOpen={isOpen}
                        toggle={toggle} />
                    <div className="row row--single row--header">
                        <div className="col-xs-12 col-md-6 user__container__body__header">
                            <div className="user__container__body__header-img-container">
                                <img
                                    className="user__container__body__header-img"
                                    src={local_user.avatar ? `http://api.terralogic.ngrok.io/${local_user.avatar}` : avatar} alt="avatar" />
                                <img
                                    data-testid="toggle-btn"
                                    onClick={toggle}
                                    className="user__container__body__header-subimg"
                                    src={editSvg}
                                    alt="edit" />
                            </div>
                            <div className="user__container__body__header-name">{local_user.displayName}</div>
                        </div>
                    </div>
                </>}
            {local_user && <UserForm avatar={avatarLink} updateProfile={updateProfile} user={local_user} logout={logoutUser} />}
        </div>
    );
}

UserPageInner.propTypes = {
    isAuth: PropTypes.bool.isRequired,
    logoutUser: PropTypes.func.isRequired,
    userLoading: PropTypes.bool.isRequired,
    uploadImage: PropTypes.func.isRequired,
    local_user: PropTypes.object.isRequired
}

export default UserPageInner
