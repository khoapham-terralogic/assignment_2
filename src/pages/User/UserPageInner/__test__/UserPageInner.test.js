import React from 'react'
import UserPageInner from '..'
import { shallow } from 'enzyme'

describe('Test <UserPageInner />', () => {
    const props = {
        isAuth: false,
        logoutUser: jest.fn(),
        updateProfile: jest.fn(),
        userLoading: false,
        uploadImage: jest.fn(),
        avatarLink: "link-to-avatar.jpg",
        local_user: {
            name: "haha",
            email: "haha@mail.com",
            phone: "0123456789",
            avatar: "another-link-to-avatar.jpg"
        }
    }
    it("should render", () => {
        const wrapper = shallow(<UserPageInner {...props} />)
        expect(wrapper).toMatchSnapshot()
    })
    it('should render loading component if isLoading is true', () => {
        const wrapper = shallow(<UserPageInner {...props} />)
        expect(wrapper.find(".loadingBackground")).toHaveLength(0)
        wrapper.setProps({ userLoading: true });
        expect(wrapper.find(".loadingBackground")).toHaveLength(1)
        expect(wrapper).toMatchSnapshot()
    });
    it('should open modal when onclick is called', () => {
        const changeState = jest.fn()
        const wrapper = shallow(<UserPageInner onClick={changeState} {...props} />)
        const handleClick = jest.spyOn(React, 'useState')
        handleClick.mockImplementation(isOpen => [isOpen, changeState]);
        wrapper.find('[data-testid="toggle-btn"]').simulate('click')
        expect(changeState).toBeTruthy()
    })
    it('should push to / if isAuth is false', () => {
        let useEffect
        const mockUseEffect = () => {
            useEffect.mockImplementationOnce(f => f());
        };
        beforeEach(() => {
            useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
            mockUseEffect();
            mockUseEffect();
            const wrapper = shallow(<UserPageInner {...props} />);
            expect(wrapper.find(UserPageInner)).toHaveLength(0)
        });
    })
    it('should render default avatar if user avatar is null', () => {
        const wrapper = shallow(<UserPageInner {...props} />)
        expect(wrapper.find('.user__container__body__header-img').prop('src')).toEqual("http://api.terralogic.ngrok.io/another-link-to-avatar.jpg");
        wrapper.setProps({ local_user: { avatar: null } })
        expect(wrapper.find('.user__container__body__header-img').prop('src')).toEqual("avatar.png")
    });
    it('should push to / if isAuth is false', () => {
        let useEffect
        const mockUseEffect = () => {
            useEffect.mockImplementationOnce(f => f());
        };
        beforeEach(() => {
            useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
            mockUseEffect();
            mockUseEffect();
            const wrapper = shallow(<UserPageInner {...props} />);
            expect(wrapper.find(UserPageInner)).toHaveLength(1)
            wrapper.setProps({ isAuth: false })
            expect(wrapper.find(UserPageInner)).toHaveLength(0)
        });
    })
})
