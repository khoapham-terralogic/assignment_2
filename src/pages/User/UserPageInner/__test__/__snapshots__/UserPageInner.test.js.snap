// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <UserPageInner /> should render 1`] = `
<div
  className="user__container__body"
>
  <CustomModal
    isLoading={false}
    isOpen={false}
    modalCallback={[Function]}
    toggle={[Function]}
    uploadImage={[MockFunction]}
  />
  <div
    className="row row--single row--header"
  >
    <div
      className="col-xs-12 col-md-6 user__container__body__header"
    >
      <div
        className="user__container__body__header-img-container"
      >
        <img
          alt="avatar"
          className="user__container__body__header-img"
          src="http://api.terralogic.ngrok.io/another-link-to-avatar.jpg"
        />
        <img
          alt="edit"
          className="user__container__body__header-subimg"
          data-testid="toggle-btn"
          onClick={[Function]}
          src="edit_photo.svg"
        />
      </div>
      <div
        className="user__container__body__header-name"
      />
    </div>
  </div>
  <WithFormik(UserInnerForm)
    avatar="link-to-avatar.jpg"
    logout={[MockFunction]}
    updateProfile={[MockFunction]}
    user={
      Object {
        "avatar": "another-link-to-avatar.jpg",
        "email": "haha@mail.com",
        "name": "haha",
        "phone": "0123456789",
      }
    }
  />
</div>
`;

exports[`Test <UserPageInner /> should render loading component if isLoading is true 1`] = `
<div
  className="user__container__body"
>
  <div
    className="loadingBackground"
  >
    <ClipSpinner
      color="#fff"
      size={45}
    />
  </div>
  <CustomModal
    isLoading={true}
    isOpen={false}
    modalCallback={[Function]}
    toggle={[Function]}
    uploadImage={[MockFunction]}
  />
  <div
    className="row row--single row--header"
  >
    <div
      className="col-xs-12 col-md-6 user__container__body__header"
    >
      <div
        className="user__container__body__header-img-container"
      >
        <img
          alt="avatar"
          className="user__container__body__header-img"
          src="http://api.terralogic.ngrok.io/another-link-to-avatar.jpg"
        />
        <img
          alt="edit"
          className="user__container__body__header-subimg"
          data-testid="toggle-btn"
          onClick={[Function]}
          src="edit_photo.svg"
        />
      </div>
      <div
        className="user__container__body__header-name"
      />
    </div>
  </div>
  <WithFormik(UserInnerForm)
    avatar="link-to-avatar.jpg"
    logout={[MockFunction]}
    updateProfile={[MockFunction]}
    user={
      Object {
        "avatar": "another-link-to-avatar.jpg",
        "email": "haha@mail.com",
        "name": "haha",
        "phone": "0123456789",
      }
    }
  />
</div>
`;
